display_list/display_list_unittests.cc

```diff
@@ -1580,5 +1616,143 @@ TEST(DisplayList, SaveLayerBoundsSnapshotsImageFilter) {
+  builder.setColor(SkColorSetARGB(127, 255, 255, 255));
+  builder.setImageFilter(TestImageFilter1);
+  builder.saveLayer(nullptr, true);
+  builder.setImageFilter(nullptr);
+  builder.drawRect({10, 10, 20, 20});
+  builder.restore();
+  builder.Build()->Dispatch(expector);
+  EXPECT_EQ(expector.save_layer_count(), 1);
+}
+
+TEST(DisplayList, SaveLayerColorFilterPreventsOpacityOptimization) {
+  SaveLayerOptions expected = SaveLayerOptions::kWithAttributes;
+  SaveLayerOptionsExpector expector(expected);
+
+  DisplayListBuilder builder;
+  builder.setColor(SkColorSetARGB(127, 255, 255, 255));
+  builder.setColorFilter(TestColorFilter1);
+  builder.saveLayer(nullptr, true);
+  builder.setColorFilter(nullptr);
+  builder.drawRect({10, 10, 20, 20});
+  builder.restore();
+  builder.Build()->Dispatch(expector);
+  EXPECT_EQ(expector.save_layer_count(), 1);
+}
+
+TEST(DisplayList, SaveLayerSrcBlendPreventsOpacityOptimization) {
+  SaveLayerOptions expected = SaveLayerOptions::kWithAttributes;
+  SaveLayerOptionsExpector expector(expected);
+
+  DisplayListBuilder builder;
+  builder.setColor(SkColorSetARGB(127, 255, 255, 255));
+  builder.setBlendMode(SkBlendMode::kSrc);
+  builder.saveLayer(nullptr, true);
+  builder.setBlendMode(SkBlendMode::kSrcOver);
+  builder.drawRect({10, 10, 20, 20});
+  builder.restore();
+  builder.Build()->Dispatch(expector);
+  EXPECT_EQ(expector.save_layer_count(), 1);
+}
+
+TEST(DisplayList, SaveLayerImageFilterOnChildSupportsOpacityOptimization) {
+  SaveLayerOptions expected =
+      SaveLayerOptions::kWithAttributes.with_can_distribute_opacity();
+  SaveLayerOptionsExpector expector(expected);
+
+  DisplayListBuilder builder;
+  builder.setColor(SkColorSetARGB(127, 255, 255, 255));
+  builder.saveLayer(nullptr, true);
+  builder.setImageFilter(TestImageFilter1);
+  builder.drawRect({10, 10, 20, 20});
+  builder.restore();
+  builder.Build()->Dispatch(expector);
+  EXPECT_EQ(expector.save_layer_count(), 1);
+}
+
+TEST(DisplayList, SaveLayerColorFilterOnChildPreventsOpacityOptimization) {
+  SaveLayerOptions expected = SaveLayerOptions::kWithAttributes;
+  SaveLayerOptionsExpector expector(expected);
+
+  DisplayListBuilder builder;
+  builder.setColor(SkColorSetARGB(127, 255, 255, 255));
+  builder.saveLayer(nullptr, true);
+  builder.setColorFilter(TestColorFilter1);
+  builder.drawRect({10, 10, 20, 20});
+  builder.restore();
+  builder.Build()->Dispatch(expector);
+  EXPECT_EQ(expector.save_layer_count(), 1);
+}
+
+TEST(DisplayList, SaveLayerSrcBlendOnChildPreventsOpacityOptimization) {
+  SaveLayerOptions expected = SaveLayerOptions::kWithAttributes;
+  SaveLayerOptionsExpector expector(expected);
+
+  DisplayListBuilder builder;
+  builder.setColor(SkColorSetARGB(127, 255, 255, 255));
+  builder.saveLayer(nullptr, true);
+  builder.setBlendMode(SkBlendMode::kSrc);
+  builder.drawRect({10, 10, 20, 20});
+  builder.restore();
+  builder.Build()->Dispatch(expector);
+  EXPECT_EQ(expector.save_layer_count(), 1);
+}
```

Comment:
I don't see any tests with nested saveLayer calls. We should probably add one that exercises this optimisation, and one that doesn't.

Reply:
I added a couple. `Can(Cannot(..., Can(....)))` and `Can(Can(...))`